pnpbridge/src/adapters/src/mqtt_pnp/json_rpc_protocol_handler.hpp

```diff
@@ -14,20 +21,35 @@ class JsonRpcProtocolHandler : public MqttProtocolHandler {
         JSON_Value*             ProtocolHandlerConfig
     );
 
-    // TODO: Handle command input
-    void
-    OnPnpMethodCall(
-        const DIGITALTWIN_CLIENT_COMMAND_REQUEST*   CommandRequest,
-        DIGITALTWIN_CLIENT_COMMAND_RESPONSE*        CommandResponse
+    static void
+    OnPnpPropertyCallback(
+        _In_ PNPBRIDGE_COMPONENT_HANDLE PnpComponentHandle,
+        _In_ const char* PropertyName,
+        _In_ JSON_Value* PropertyValue,
+        _In_ int version,
+        _In_ void* userContextCallback
+    );
+
+    static int
+    OnPnpCommandCallback(
+        _In_ PNPBRIDGE_COMPONENT_HANDLE PnpComponentHandle,
+        _In_ const char* CommandName,
+        _In_ JSON_Value* CommandValue,
+        _Out_ unsigned char** CommandResponse,
+        _Out_ size_t* CommandResponseSize
     );

```

Comment:
imo statics should be defined as part of MqttProtocolHandler interface, then they should call into MqttProtocolHandler handler functions on the instance, which will actually be implemented by the handlers e.g. JsonRpcProtocolHandler

Reply:
actually I think these static handlers can just be lambdas within mqtt_pnp.cpp

---
In reply to: [469537393](https://github.com/Azure/AzurePnPBridgePreview/pull/34#discussion_r469537393) [](ancestors = 469537393)

Reply:
I'm wondering how this would work with polymorphism since both methods have to be static and MqttProtocolHandler is a pure abstract class. Let's discuss offline.

---
In reply to: [469537393](https://github.com/Azure/AzurePnPBridgePreview/pull/34#discussion_r469537393) [](ancestors = 469537393)